XRDDEV-1725 update OpenApiGenerator

Dos4dev opened the pull request.

niis-jenkins commented:
Can one of the admins verify this patch?

jhyoty reviewed (changes requested):
The build fails with the following error:
```
> Task :common-rest-api:compileJava FAILED
.../X-Road/src/common-rest-api/src/main/java/org/niis/xroad/restapi/exceptions/ApplicationExceptionHandler.java:32: error: package org.niis.xroad.restapi.openapi.model does not exist
import org.niis.xroad.restapi.openapi.model.ErrorInfo;                                           
```
It seems that adding the "apiFirst: true" additional option disables model and api code generation. Disabling the option causes another error:
```
> Task :common-rest-api:compileJava FAILED
.../X-Road/src/common-rest-api/src/main/java/org/niis/xroad/restapi/exceptions/ExceptionTranslator.java:104: error: incompatible types: List<String> cannot be converted to Set<String>
                result.setMetadata(metadata);
```
Which is caused by the newer generator using Set instead of List if the api uses `uniqueItems: true` property. The code should be refactored to use Set instead. 

Likely the proxy-ui-api module has similar issues.

Dos4dev commented:
> The build fails with the following error:
> 
> ```
> > Task :common-rest-api:compileJava FAILED
> .../X-Road/src/common-rest-api/src/main/java/org/niis/xroad/restapi/exceptions/ApplicationExceptionHandler.java:32: error: package org.niis.xroad.restapi.openapi.model does not exist
> import org.niis.xroad.restapi.openapi.model.ErrorInfo;                                           
> ```
> 
> It seems that adding the "apiFirst: true" additional option disables model and api code generation. Disabling the option causes another error:
> 
> ```
> > Task :common-rest-api:compileJava FAILED
> .../X-Road/src/common-rest-api/src/main/java/org/niis/xroad/restapi/exceptions/ExceptionTranslator.java:104: error: incompatible types: List<String> cannot be converted to Set<String>
>                 result.setMetadata(metadata);
> ```
> 
> Which is caused by the newer generator using Set instead of List if the api uses `uniqueItems: true` property. The code should be refactored to use Set instead.
> 
> Likely the proxy-ui-api module has similar issues.

I added **"apiFirst: true"**  properties because I received another problem after refactored "from List to Set". 
Generated code has not valid import 
`springfox.documentation.annotations.ApiIgnore`
and compiler fails on this import. 
I added **"apiFirst: true"**, to fix this problem,  but I am not sure that is the best solution. 
Similar problem [link](https://github.com/OpenAPITools/openapi-generator/issues/8192)
I used to be stuck with running Gradle build task, but now I have progress, so it would be easy to find a better solution. I will update this PR soon. 
@jhyoty  thank you for the reply.

jhyoty commented:
`io.springfox:springfox-core:3.0.0` seems to include the missing annotation, so one solution is to add that as a dependency. 
The intention is to use "contract first" process where model classes and API interfaces are generated from the OpenAPI spec. If "apiFirst: true" disables that, it should not be used.

jhyoty commented:
The build still fails:
```
./gradlew clean
./gradlew build
...
> Task :common-rest-api:compileJava FAILED
.../X-Road/src/common-rest-api/src/main/java/org/niis/xroad/restapi/exceptions/ApplicationExceptionHandler.java:32: error: package org.niis.xroad.restapi.openapi.model does not exist
import org.niis.xroad.restapi.openapi.model.ErrorInfo;
...
```
There also seem to be some changes to common-verifier module, and a test failure:
```
> Task :common-verifier:test
ee.ria.xroad.common.conf.globalconf.GlobalConfTest > getInstanceIdentifiers FAILED
    java.lang.AssertionError at GlobalConfTest.java:135
```
It would be best to keep the changes in proxy-ui-api and common-rest-api modules and not touch other parts of the system.

Dos4dev commented:
> common-verifier module

I made changes in common-verifier module because they related to other modules.

> The build still fails:
> 
> ```
> ./gradlew clean
> ./gradlew build
> ...
> > Task :common-rest-api:compileJava FAILED
> .../X-Road/src/common-rest-api/src/main/java/org/niis/xroad/restapi/exceptions/ApplicationExceptionHandler.java:32: error: package org.niis.xroad.restapi.openapi.model does not exist
> import org.niis.xroad.restapi.openapi.model.ErrorInfo;
> ...
> ```
> 
> There also seem to be some changes to common-verifier module, and a test failure:
> 
> ```
> > Task :common-verifier:test
> ee.ria.xroad.common.conf.globalconf.GlobalConfTest > getInstanceIdentifiers FAILED
>     java.lang.AssertionError at GlobalConfTest.java:135
> ```
> 
> It would be best to keep the changes in proxy-ui-api and common-rest-api modules and not touch other parts of the system.

I removed changes from common-verifier module, and now build can be created successfully. @jhyoty could you please take a look? 
What do you think if we squash all commits from the 1725 branch into one before merge to the developer branch?

jhyoty reviewed (changes requested):
Some minor issues, but otherwise looks good to me.

jhyoty commented:
> What do you think if we squash all commits from the 1725 branch into one before merge to the developer branch?
Yes, it would be clearer if you could rebase on develop and  squash the commits.

Dos4dev commented on src/proxy-ui-api/src/main/java/org/niis/xroad/securityserver/restapi/openapi/SecurityServersApiController.java:
```diff
@@ -73,13 +75,13 @@
             SecurityServerId currentSecurityServerId = serverConfService.getSecurityServerId();
             SecurityServer currentSecurityServer = securityServerConverter.convert(currentSecurityServerId);
-            securityServers = Collections.singletonList(currentSecurityServer);
+            securityServers = new HashSet<>(Collections.singletonList(currentSecurityServer));
```
I have to use Set. 
Because generated interface **SecurityServersApi** has a method getSecurityServers() with response type `ResponseEntity<Set<SecurityServer>> `. I can try to change OpenApiGenerator configuration to change response from Set to List?

jhyoty replied:
You can simplify ` new HashSet<>(Collections.singletonList(currentSecurityServer))` to `Collections.singleton(currentSecurityServer)` which is a Set.

See https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#singleton-T-

Dos4dev replied:
fixed in https://github.com/nordic-institute/X-Road/pull/1042/commits/c7d8fa152ba6d424760c35a3067ce80f4b687612

Dos4dev commented on src/proxy-ui-api/src/main/resources/openapi-definition.yaml:
```diff
@@ -5148,8 +5148,6 @@ components:
           description: access right given at
           example: '2018-12-15T00:00:00.001Z'
           readOnly: true
-          minLength: 24
-          maxLength: 24
```
I removed it because it creates next exception.
```
HV000030: No validator could be found for constraint 'javax.validation.constraints.Size' validating type 'java.time.OffsetDateTime'. Check configuration for 'addServiceClientAccessRights.accessRights.items[].rightsGivenAt'
javax.validation.UnexpectedTypeException: HV000030: No validator could be found for constraint 'javax.validation.constraints.Size' validating type 'java.time.OffsetDateTime'. Check configuration for 'addServiceClientAccessRights.accessRights.items[].rightsGivenAt'
```

jhyoty replied:
Makes sense. Maybe the old generator did not even try to add the constraint annotations.

Dos4dev commented on src/proxy-ui-api/src/test/java/org/niis/xroad/securityserver/restapi/auth/CsrfWebMvcTest.java:
```diff
@@ -194,8 +196,9 @@ public void getUserCsrfCookieFail() throws Exception {
                 .username(username)
-                .roles(Collections.singletonList(Role.XROAD_SECURITYSERVER_OBSERVER.getGrantedAuthorityName()))
-                .permissions(userPermissions);
+                .roles(new HashSet<>(Collections.singletonList(
```
Fixed in https://github.com/nordic-institute/X-Road/pull/1042/commits/c7d8fa152ba6d424760c35a3067ce80f4b687612

Dos4dev commented on src/proxy-ui-api/src/test/java/org/niis/xroad/securityserver/restapi/auth/CsrfWebMvcTest.java:
```diff
@@ -137,8 +138,9 @@ public void login() throws Exception {
                 .username(username)
-                .roles(Collections.singletonList(Role.XROAD_SECURITYSERVER_OBSERVER.getGrantedAuthorityName()))
-                .permissions(userPermissions);
+                .roles(new HashSet<>(Collections.singletonList(
```
Fixed in https://github.com/nordic-institute/X-Road/pull/1042/commits/c7d8fa152ba6d424760c35a3067ce80f4b687612

Dos4dev commented:
Jarkko, than you for the help :bowing_man:
